File: Permutation in String - Leetcode 567/Permutation in String - Leetcode 567.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        n1 = len(s1)
        n2 = len(s2)

        if n1 > n2:
            return False

        s1_counts = [0] * 26
        s2_counts = [0] * 26

        for i in range(n1):
            s1_counts[ord(s1[i]) - 97] += 1
            s2_counts[ord(s2[i]) - 97] += 1

        if s1_counts == s2_counts:
            return True

        for i in range(n1, n2):
            s2_counts[ord(s2[i]) - 97] += 1
            s2_counts[ord(s2[i - n1]) - ord("a")] -= 1
            if s1_counts == s2_counts:
                return True

        return False

# Time Complexity: O(n)
# Space Complexity: O(1)

File: Permutation in String - Leetcode 567/Permutation in String - Leetcode 567.js
function checkInclusion(s1, s2) {
    const n1 = s1.length;
    const n2 = s2.length;

    if (n1 > n2) return false;

    const s1Counts = Array(26).fill(0);
    const s2Counts = Array(26).fill(0);

    for (let i = 0; i < n1; i++) {
        s1Counts[s1.charCodeAt(i) - 97]++;
        s2Counts[s2.charCodeAt(i) - 97]++;
    }

    if (s1Counts.every((val, index) => val === s2Counts[index])) return true;

    for (let i = n1; i < n2; i++) {
        s2Counts[s2.charCodeAt(i) - 97]++;
        s2Counts[s2.charCodeAt(i - n1) - 97]--;
        if (s1Counts.every((val, index) => val === s2Counts[index])) return true;
    }

    return false;
}
